### PR TITLE
Changelog 1.0.0

### PR DESCRIPTION
**Pull request by jrfnl**
* Merges in `master` to pull in the changelog from the `0.14.1` release and to fix the imaginary merge conflict in PR #1392 
* Adds the changelog for the `1.0.0` release. And yes, it's massive..... careful scrutiny appreciated.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -2,12 +2,199 @@
...
+
+Please read the complete changelog carefully before you upgrade.
+
+If you are a maintainer of an external standard based on WPCS and any of your custom sniffs are based on or extend WPCS sniffs, the same applies.
+
+### Added
+- `WordPress.PHP.PregQuoteDelimiter` sniff to the `WordPress-Extra` ruleset to warn about calls to `preg_quote()` which don't pass the `$delimiter` parameter.
+- `WordPress.Security.SafeRedirect` sniff to the `WordPress-Extra` ruleset to warn about potential open redirect vulnerabilities.
+- `WordPress.WP.DeprecatedParameterValues` sniff to the `WordPress-Extra` ruleset to detect deprecated parameter values being passed to select functions.
+- `WordPress.WP.EnqueuedResourceParameters` sniff to the `WordPress-Extra` ruleset to detect calls to the script/style register/enqueue functions which don't pass a `$version` for the script/style, which can cause issues with browser caching, and/or calls to the register/enqueue script functions which don't pass the `$in_footer` parameter, which causes scripts - by default - to be loaded in the HTML header in a layout rendering blocking manner.
```
Comment (GaryJones) [inline CHANGELOG.md]: Veeeeery long sentence!

File: CHANGELOG.md [hunk 2]
```diff
@@ -2,12 +2,199 @@
...
+- Assignments in conditions in ternaries as detected by the `WordPress.CodeAnalysis.AssignmentInCondition` sniff will now be reported under a separate error code `FoundInTernaryCondition`.
+- The default error level for the notices from the `WordPress.DB.DirectDatabaseQuery` sniff has been lowered from `error` to `warning`. For the `VIP` ruleset, the error level remains `error`.
+- The default error level for the notices from the `WordPress.Security.PluginMenuSlug` sniff has been lowered from `error` to `warning`. For the `VIP` ruleset, the error level remains `error`.
+- The default error level for the notices from the `WordPress.WP.CronInterval` sniff has been lowered from `error` to `warning`. For the `VIP` ruleset, the error level remains `error`.
+- The `Sniff::get_function_call_parameters()` utility method now has improved handling of closures when passed as function call parameters.
+- Rulesets: a number of error codes were previously silenced by explicitely `exclude`-ing them. Now, they will be silenced by setting the `severity` to `0` which makes it more easily discoverable for maintainers of custom rulesets how to enable these error codes again.
```
Comment (GaryJones) [inline CHANGELOG.md]: Typo: explicitely -> explicitly

File: CHANGELOG.md [hunk 3]
```diff
@@ -2,12 +2,199 @@
...
+    - `WordPress.VIP.SessionFunctionsUsage` without replacement.
+    - `WordPress.VIP.SessionVariableUsage` without replacement.
+    - `WordPress.VIP.SuperGlobalInputUsage` without replacement.
+- The following sniff categories have been deprecated and will be removed in WPCS 2.0.0:
+    - `CSRF`
+    - `Variables`
+    - `XSS`
+- The `posts_per_page` property in the `WordPress.VIP.PostsPerPage` sniff has been deprecated as the related functionality has been moved to the `WordPress.WP.PostsPerPage` sniff.
+    See [WP PostsPerPage: post limit](https://github.com/WordPress-Coding-Standards/WordPress-Coding-Standards/wiki/Customizable-sniff-properties#wp-postsperpage-post-limit) for more information about this property.
+- The `exclude` property which is available to most sniffs which extend the `AbstractArrayAssignmentRestrictions`, `AbstractFunctionRestrictions` and `AbstractVariableRestrictions` classes or any of their children, used to be a `string` property and expected a comma-delimited list of groups to exclude.
```
Comment (GaryJones) [inline CHANGELOG.md]: Did this require a minimum version of PHPCS?

File: CHANGELOG.md (hunk 2 above)
Comment (jrfnl) [reply CHANGELOG.md]: Why do they make English such a difficult language to spell ? ;-)

File: CHANGELOG.md (hunk 3 above)
Comment (jrfnl) [reply CHANGELOG.md]: No, it doesn't. It just makes it more forward-compatible and allows the PHPCS 3.3.0 XML arrays to be used in a custom ruleset, but the change in itself does not *need* PHPCS 3.3.0.

**Comment by jrfnl** [comment]
I've processed @GaryJones' feedback and added the changes in a separate commit. Commits should be squashed on (or before) merge.

File: CHANGELOG.md (hunk 2 above)
Comment (GaryJones) [reply CHANGELOG.md]: Oddly enough, this is one of the words that always use to be a gotcha for me, which is why I now look out for it :-)

Comment (jrfnl) [reply CHANGELOG.md]: *grin*, well, I think I have a number of those, though `(PHP)compatibility` is no longer amongst them ;-)

File: CHANGELOG.md [hunk 4]
```diff
@@ -2,12 +2,201 @@
...
+    - `WordPress.Arrays.ArrayDeclarationSpacing`: the fixer will now handle the new style annotations correctly.
+    - `WordPress.Arrays.CommaAfterArrayItem`: prevent a fixer loop when new style annotations are encountered.
+    - `WordPress.Files.FileName`: respect the new style annotations if these would selectively disable this sniff.
+    - `WordPress.WhiteSpace.ControlStructureSpacing`: handle the new style annotations correctly for the "blank line at the start/end of control structure" checks and prevent a fixer conflict when the new style annotations are encountered.
+    - `WordPress.WhiteSpace.PrecisionAlignment`: allow for checking of for precision alignment on lines containing new style annotations when `phpcs` is run with `--ignore-annotations`.
+- The `Sniff::is_test_class()` method now has improved recognition of namespaced test classes. This positively affects the following sniffs `WordPress.Files.FileName`, `WordPress.NamingConventions.PrefixAllGlobals` and `WordPress.WP.GlobalVariablesOverride` which each allow for test classes to (partially) not comply with the rules these sniffs check for.
```
Comment (JDGrimes) [inline CHANGELOG.md]: Seems like there should be some punctuation around the sniff names here, or the sentence should be reworded.

File: CHANGELOG.md [hunk 5]
```diff
@@ -2,12 +2,201 @@
...
+    - The sniff will now correctly ignore the open and close brackets of short arrays.
+    - The sniff would throw false positives when `echo`, `print`, `exit` or `die` were encountered as constants, function or class names. While it may not be a good idea to use PHP keywords in such a way, it is allowed, so the sniff should handle this correctly.
+- The `WordPress.WhiteSpace.ControlStructureSpacing` sniff would inadvertly throw an error for the spacing around the colon for a return type in a function declaration.
+- The `WordPress.WP.AlternativeFunctions` sniff used to flag all function calls to `file_get_contents()` twice, suggesting to use `wp_remote_get()` - which is only applicable for remote URLs - and the `WP_FileSystem` API - which is not needed when just _reading_ local files. These messages contradicted each other.
+    The sniff will now try to determine whether the file requested is local or remote and will only throw a `warning` suggesting to use `wp_remote_get()`, if a remote URL is being requested or when it could not be determined if the requested file is locale or remote.
```
Comment (JDGrimes) [inline CHANGELOG.md]: locale -> local

File: CHANGELOG.md [hunk 6]
```diff
@@ -2,12 +2,201 @@
 
 All notable changes to this project will be documented in this file.
 
-This projects adheres to [Semantic Versioning](http://semver.org/) and [Keep a CHANGELOG](http://keepachangelog.com/).
+This projects adheres to [Semantic Versioning](https://semver.org/) and [Keep a CHANGELOG](https://keepachangelog.com/).
 
 ## [Unreleased]
 
 _Nothing yet._
```
[the remark this answers is not in the document]
Comment (jrfnl) [reply CHANGELOG.md]: I think I prefer @JDGrimes' suggestion. Having the section there at least indicates that if people download `develop` they should be aware that there may be undocumented changes included.
Removing it completely on release and bringing it back as as soon as the release is done is the alternative as that's the only moment when there are no undocumented changes.

[the remark this answers is not in the document]
Comment (jrfnl) [reply CHANGELOG.md]: I've removed it for the release. I'll send in a PR to bring it back once the release is done.

**Comment by jrfnl** [comment]
I've processed @JDGrimes' feedback and added the changes in a separate commit. Commits should be squashed on (or before) merge.

File: CHANGELOG.md [hunk 7]
```diff
@@ -2,11 +2,197 @@
...
+- New utility method `Sniff::is_use_of_global_constant()`.
+- A rationale to the package suggestion made via `composer.json`.
+- CI: Validation of the `composer.json` file on each build.
+- A wiki page with instructions on how to [set up WPCS to run with Eclipse on XAMPP](https://github.com/WordPress-Coding-Standards/WordPress-Coding-Standards/wiki/How-to-use-WPCS-with-Eclipse-and-XAMPP).
+- Readme: A link to an external resource with more examples for setting up PHPCS for CI.
+- Readme: A badge-based quick overview of the project.
+
+### Changed
+- The `WordPress` ruleset no longer includes the `WordPress-VIP` ruleset, nor does it include any of the (deprecated) `VIP` sniffs anymore.
+- The following sniffs have been been moved to a new category:
```
Comment (Ismail-elkorchi) [inline CHANGELOG.md]: The word been seems to be unintentionally repeated.

File: CHANGELOG.md [hunk 8]
```diff
@@ -2,11 +2,197 @@
...
+- The `WordPress.VIP.PostsPerPage` sniff has been split into two distinct sniffs:
+    - `WordPress.WP.PostsPerPage` which will check for the use of a high pagination limit and will throw a `warning` when this is encountered. For the `VIP` ruleset, the error level remains `error`.
+    - `WordPress.VIP.PostsPerPage` wich will check for disabling of pagination.
+- The default value for `minimum_supported_wp_version`, as used by a [number of sniffs detecting usage of deprecated WP features](https://github.com/WordPress-Coding-Standards/WordPress-Coding-Standards/wiki/Customizable-sniff-properties#minimum-wp-version-to-check-for-usage-of-deprecated-functions-classes-and-function-parameters), has been updated to `4.6`.
+- The `WordPress.WP.AlternativeFunctions` sniff will now only throw a warning if/when the recommended alternative function is available in the minimum supported WP version of a project.
+    In addition to this, certain alternatives are only valid alternatives in certain circumstances, like when the WP version only supports the first parameter of the PHP function it is trying to replace.
+    This will will now be taken into account for:
```
Comment (Ismail-elkorchi) [inline CHANGELOG.md]: The word will seems to be unintentionally repeated.

File: CHANGELOG.md [hunk 9]
```diff
@@ -2,11 +2,197 @@
...
+- Various performance optimizations which should most notably make a difference when running WPCS on PHP 7.
+- References to the WordPress.com VIP platform have been clarified.
+- Unit Tests: custom properties set in unit test files are reset after use.
+- Various improvements to the ruleset used by the WPCS project itself and minor code clean up related to this.
+- CI: Each change will now also be tested against the lowest supported PHPCS 3 version.
+- CI: Each change will now also be checked for PHP cross-version compatibility.
+- CI: Minor changes to the script to make the build testing faster.
+- Updated the [custom ruleset example](https://github.com/WordPress-Coding-Standards/WordPress-Coding-Standards/blob/develop/phpcs.xml.dist.sample) for the changes contained in this release and to reflect current best practices regarding the PHPCompatibility standard.
+- The instructions on how to set up WPCS for various IDEs have been moved from the `README` to the [wiki](https://github.com/WordPress-Coding-Standards/WordPress-Coding-Standards/wiki).
+- Updated output examples in `README.md` and `CONTRIBUTING.md` and minor other changes to these files.
```
Comment (Ismail-elkorchi) [inline CHANGELOG.md]: The order of the adjectives in this noun phrase is unusual. minor other => other minor.

File: CHANGELOG.md [hunk 10]
```diff
@@ -2,11 +2,197 @@
...
+    - The `custom_test_class_whitelist` property which is available for each of these sniffs expects to be passed a Fully Qualified Class Name. FQCNs prefixed with a global namespace indicator will now be correctly handled.
+- The determination of whether a `T_STRING` is a function call or not has been improved in the `AbstractFunctionRestrictions` class. This improvement benefits all sniffs which extend this abstract and any of its children (> 10 sniffs) and fixes the following false positives:
+    - Class declarations will no longer be confused with function calls.
+    - Use statement alias declarations will no longer be confused with function calls.
+- Various bugs in the `WordPress.Arrays.ArrayIndentation` sniff:
+    - The sniff will no longer throw false positives or try to fix multi-line text strings where the closing quote is on a line by itself.
+    - The sniff would go into a fixer loop when it encountered a multi-line trailing comment after an array item.
+- The `WordPress.CodeAnalysis.AssignmentInCondition` was throwing false positives for ternaries in nested, but unrelated, parenthesis.
```
Comment (Ismail-elkorchi) [inline CHANGELOG.md]: Parentheses should be plural since it does not have a determiner and is countable.

File: CHANGELOG.md [hunk 11]
```diff
@@ -2,11 +2,197 @@
...
+    - Parameters in a closure declaration were incorrectly being regarded as global variables.
+    - Non-prefixed variables created by a `foreach()` construct in the global namespace were previously not detected.
+    - Non-prefixed globals found in namespaced test classes should be ignored by the sniff, but were not.
+    - Definition of non-prefixed global WP constants which are intended to be overruled, should not trigger an error from this sniff.
+    - The sniff presumed the WP naming conventions for PHP constructs, while it should check for the construct being prefixed regardless of whether camelCase, PascalCase, snake_case or other naming conventions are used.
+    - The sniff presumed the WP naming conventions for prefixes used in hook names. The sniff will now be more tolerant when non-conventional word separators are used in prefixes for hooks.
+- The `WordPress.NamingConventions.ValidFunctionName` sniff no longer "hides" one message behind another. The sniff will now correctly throw a message about function names not being in `snake_case`, even when the `FunctionDoubleUnderscore` or `MethodDoubleUnderscore` errorcodes have been excluded.
```
Comment (Ismail-elkorchi) [inline CHANGELOG.md]: Errorcodes should not be written as a single word.

File: CHANGELOG.md [hunk 12]
```diff
@@ -2,11 +2,197 @@
...
+- The `WordPress.PHP.StrictInArray` sniff will no longer throw an error when `in_array`, `array_search` or `array_keys` are used in a file `use` statement.
+- Various bugs in the `WordPress.Security.EscapeOutput` sniff:
+    - A limited list of native PHP constants which are safe to use, such as `PHP_EOL`, has been added. When any of these constants are encountered, the sniff will no longer demand output escaping for them.
+    - The sniff was underreporting issues with variables passed to `trigger_error()`.
+    - While reporting an issue, sometimes the wrong error message was used. The sniff logic has been adjusted to prevent this.
+    - The sniff will now correctly ignore the open and close brackets of short arrays.
+    - The sniff would throw false positives when `echo`, `print`, `exit` or `die` were encountered as constants, function or class names. While it may not be a good idea to use PHP keywords in such a way, it is allowed, so the sniff should handle this correctly.
+- The `WordPress.WhiteSpace.ControlStructureSpacing` sniff would inadvertly throw an error for the spacing around the colon for a return type in a function declaration.
```
Comment (Ismail-elkorchi) [inline CHANGELOG.md]: The word inadvertly seems misspelled. inadvertly => inadvertently

File: CHANGELOG.md [hunk 13]
```diff
@@ -2,11 +2,197 @@
...
+- The `WordPress.WP.AlternativeFunctions` sniff used to flag all function calls to `file_get_contents()` twice, suggesting to use `wp_remote_get()` - which is only applicable for remote URLs - and the `WP_FileSystem` API - which is not needed when just _reading_ local files. These messages contradicted each other.
+    The sniff will now try to determine whether the file requested is local or remote and will only throw a `warning` suggesting to use `wp_remote_get()`, if a remote URL is being requested or when it could not be determined if the requested file is local or remote.
+- The expected default value for `wp_upload_bits()` in the `WordPress.WP.DeprecatedParameters` sniff.
+- The `WordPress.WP.GlobalVariablesOverride` sniff previously did not detect variables created by a `foreach()` construct which would override WP global variables.
+- Various bugs in the `WordPress.WP.I18n` sniff:
+    - The sniff will no longer throw false positives for calls to methods carrying the same name as any of the global WP functions being targetted and has improved handling of parse errors and live coding.
```
Comment (Ismail-elkorchi) [inline CHANGELOG.md]: To form the past participle of the verb target, the final consonant should not be doubled before adding -⁠ed. targetted => targeted.

**Comment by jrfnl** [comment]
Thanks @Ismail-elkorchi for the review. I've processed your feedback and added the changes in a separate commit.

N.B.: Commits should be squashed on (or before) merge.

**Comment by jrfnl** [comment]
Anyone up for giving the second approval for this PR ? @JDGrimes ?

No need to merge, I'll squash and merge when doing the release.

**Comment by jrfnl** [comment]
Rebased & squashed. Also added a line about the newly merged PR #1430 in the _Changed_ section:
`- CI: The rulesets will now also be tested on each change to ensure no unexpected messages are thrown.`

Once the builds have passed I'll merge this.

**Pull request merged**